### ISSUE TITLE
Adding CMake Support

### ISSUE THREAD

**Issue by siy** [modm-io/modm#10]
I've tried to follow instructions and downloaded/installed everything according to documentation. Nevertheless, when I'm trying to launch _lbuild discover-repository_ command inside _modm/examples/stm32f103c8t6_blue_pill/blink_, I'm getting following error:

> ERROR in '/home/siy/CLionProjects/modm'
> Traceback (most recent call last):
>   File "/home/siy/CLionProjects/modm/repo.lb", line 21, in <module>
>     import modm.pkg
> ImportError: No module named 'modm'

Also, there is almost no information on how to configure and build project with CMake. I'd like to extend this part of documentation, but I need some starting point, i.e. working build process (either with SCons or CMake).

**Comment by salkinium** [comment]
Did you clone the repository recursively? We have a few submodule in `/ext`, and `modm.pkg` is in the `modm/ext/modm-devices` submodule. You can do `git submodule update --init`. (~~I forgot to add this step to the docs, fixing that now~~ fixed).

> Also, there is almost no information on how to configure and build project with CMake.
> I'd like to extend this part of documentation

Yes, please! I haven't gotten to documenting the [build system generators](http://modm.io/reference/build-systems/) yet, but you can look at the [SCons generator, it should give you a good template for porting to CMake](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/scons/module.lb#L72-L142).
There is even a template for [wrapping SCons into CMake](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/cmake/CMakeLists.txt.in) ([Background](https://github.com/roboterclubaachen/xpcc/pull/344)), but I haven't ported that to lbuild either yet.

(See #9 for my documentation PR).

**Comment by siy** [comment]
Thanks for the suggestion, retrieving submodules helped.
Frankly speaking I was expecting that CMake support is already working, only documentation is missing. Nevertheless, I've started writing module for CMake support. From my point of view, there is no big sense to wrap scons build into CMake build. Instead I'll try to implement generation of full CMake-based project (so it will not depend on scons anymore). This should enable full support of CLion IDE. If also to write a small makefile which will wrap CMake build, then resulting project will be compatible with Eclipse-based IDE's. Taking into account recent purchase of Atollic (which provides Eclipse-based IDE) by ST, such a support might be useful for wider acceptance of modm.

**Comment by salkinium** [comment]
Absolutely love your efforts, since I don't have much experience with IDEs or CMake.

> From my point of view, there is no big sense to wrap scons build into CMake build.

In xpcc the code generation was implemented in SCons, so you had to call SCons anyway when using CMake, so the wrapper made sense for us. This is one reason why we made created modm, to get rid of this dependency for exactly the use-cases you describe 😄

> Frankly speaking I was expecting that CMake support is already working, only documentation is missing.

Sorry, it's just me and @dergraaf working on this at the moment, the rest of the team is still using xpcc.
I plan to switch them over this summer, then we have more devs available for improvements like these.

For compiler settings, code for tools, how they work, see [our SCons tools repo](https://github.com/modm-io/scons-build-tools/tree/master/site_tools). It would be nice to have similar behavior, maybe we can later even share for example the compiler flags between CMake and SCons. But the SCons settings are by no means the best, so whatever you think is better we'll use.

If you build modm into a static library and then link it with the application, you need to wrap the library with `--whole-archive`. [See here for explanation](https://github.com/roboterclubaachen/xpcc/pull/351#issuecomment-381364481).

**Comment by siy** [comment]
Thanks a lot for comments and suggestions.

**Comment by siy** [comment]
I'm working on the CMake build script generator. Initial implementation going to be limited to STM32 targets only, but later I plan extend it to cover all remaining targets. So far I've reached the point where build of main project binary is almost passing. Remaining issue is that link step is failing because of missing symbol:

```/usr/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/lib/thumb/v7-m/libg_nano.a(lib_a-init.o): In function `__libc_init_array':
init.c:(.text.__libc_init_array+0x10): undefined reference to `_init'```

The full linker command line looks so:

> /usr/bin/arm-none-eabi-g++       -mcpu=cortex-m3     -mthumb     -mthumb-interwork     -fdata-sections     
> -ffunction-sections     -finline-limit=10000     -fno-move-loop-invariants     -fno-split-wide-types     
> -fno-tree-loop-optimize     -fno-unwind-tables     -fshort-wchar     -funsigned-bitfields     
> -funsigned-char     -gdwarf-2     -W     -Wall     -Wextra     -Wformat     -Winit-self     -Wno-long-long     
> -Wpointer-arith     -Wundef     -Wuninitialized     -Wunused     -Wunused-parameter     -Wwrite-strings
>     -fno-exceptions     -fno-rtti     -fno-threadsafe-statics     -fuse-cxa-atexit     -Woverloaded-virtual     
> -std=c++14 -Og  --specs=nano.specs --specs=nosys.specs     -Llink     -nostartfiles     
> -T/home/siy/CLionProjects/modm/build/stm32f103c8t6_blue_pill/logger/linkerscript.ld     
> -Wl,--fatal-warnings     -Wl,--gc-sections     -Wl,--no-wchar-size-warning     -Wl,--relax     
> -Wl,-wrap,_calloc_r     -Wl,-wrap,_free_r     -Wl,-wrap,_malloc_r     
> -Wl,-wrap,_realloc_r 
> CMakeFiles/logger.dir/ext/cmsis/device/peripherals.c.o 
> CMakeFiles/logger.dir/src/modm/architecture/driver/atomic/flag.cpp.o 
> CMakeFiles/logger.dir/src/modm/board/board.cpp.o 
> CMakeFiles/logger.dir/src/modm/debug/error_report.cpp.o 
> CMakeFiles/logger.dir/src/modm/io/iodevice.cpp.o 
> CMakeFiles/logger.dir/src/modm/io/iostream.cpp.o 
> CMakeFiles/logger.dir/src/modm/io/iostream_float.cpp.o 
> CMakeFiles/logger.dir/src/modm/io/iostream_printf.cpp.o 
> CMakeFiles/logger.dir/src/modm/math/utils/bit_operation.cpp.o 
> CMakeFiles/logger.dir/src/modm/math/utils/pc/operator.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/clock/clock.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/clock/delay.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/clock/latency.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/clock/modm_clock.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/clock/systick_timer.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/core/assert.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/core/cxxabi.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/core/heap_newlib.c.o 
> CMakeFiles/logger.dir/src/modm/platform/core/heap_table.c.o 
> CMakeFiles/logger.dir/src/modm/platform/core/startup.c.o 
> CMakeFiles/logger.dir/src/modm/platform/core/syscalls.c.o 
> CMakeFiles/logger.dir/src/modm/platform/core/vectors.c.o 
> CMakeFiles/logger.dir/src/modm/platform/gpio/enable.cpp.o 
> CMakeFiles/logger.dir/src/modm/platform/uart/uart_2.cpp.o 
> CMakeFiles/logger.dir/src/modm/utils/dummy.cpp.o 
> CMakeFiles/logger.dir/main.cpp.o  -o logger.elf  

I'd appreciate any hints/comments/suggestions in regard to this issue.

**Comment by salkinium** [comment]
I think you're missing the assembly files in the core, specifically [`default_handler.sx` which stubs `_init`](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/cortex/default_handler.sx#L30-L39).

**Comment by siy** [comment]
Thanks a lot! Will look deeper into it during weekend. Once this part will be working, I'll prepare the PR.

Also, recently I took a closer look into the project file (.ioc) created by CubeMX. Looks like it holds enough information to generate board initialization code automatically. The only thing which might be necessary (rather for convenience than absolutely mandatory) is to add user labels to configured pins (and they will become names of pins in generated code). All other relevant information is present (including details of clock configuration). It also seems possible to configure other peripherals. I think that adding such a generator might be helpful to enable integration of modm with CubeMX. Overall combination of CubeMX + modm + (Clion or Eclipse or Eclipse-based) looks very attractive and may significantly increase interest to modm.
Once I finish CMake generator I'm going to take deeper look into this.

**Comment by salkinium** [comment]
Yes! That‘s what I thought as well, when I looked at those files!
A generator for our board files would be fantastic!
I‘ve written [a post about the modm device file generator](http://blog.salkinium.com/modm-devices/), maybe it could be integrated into that?

**Comment by siy** [comment]
Perhaps. At least information from CubeMX project file accompanied with already collected information about devices enables a lot of possibilities for code generation. Lets eat this elephant piece by piece and discuss details later, when I finish CMake generator.

**Comment by siy** [comment]
Meanwhile I started working on CubeMX integration. So far my idea is to add one more command to library builder - import. This command should generate project.xml (if it des not exist yet) and board-specific files (including init code and OpenOCD config). Also, with additional command line option it shoul be possible to invoke 'build' command once generation stage is finished.

P.S. are there any convenient way to communicate with modm team in regard to development? Some kind of online chat might be much more convenient.

**Comment by salkinium** [comment]
> P.S. are there any convenient way to communicate with modm team in regard to development? Some kind of online chat might be much more convenient.

I've invited you into the contributors team, then we can use the discussions forum there.
I think that's probably the best tool for discussing modm stuff, since all the GitHub specific things renders/links correctly.
https://github.com/orgs/modm-io/teams/contributors/discussions

> So far my idea is to add one more command to library builder - import

I'd make it a separate tool in modm, since lbuild is meant to be completely separate from modm (as it's used by @dergraaf for other things) and doesn't include anything specific to microcontrollers at all.

**Comment by salkinium** [comment]
I’m closing this as CMake support has been implemented in #15.

**Comment by salkinium** [comment]
I added [CMake docs to the homepage](http://modm.io/reference/build-systems/) some time ago too.
